### ISSUE TITLE
Performance Regression in TSBS Query Time from v0.9 to v0.10

### ISSUE THREAD

**Issue by discord9** [GreptimeTeam/greptimedb#5052]
### What type of enhancement is this?

Performance

### What does the enhancement do?

__edited__: could be false alarm since the old 0.9 tsbs tests is on a different machine(although with similar spec), still working to confirm if test on the same machine can reproduce this
__edited 2__: add v0.9.5 and v0.9.4 tsbs(on same machine) results, 
__edited 3__: after compare more version, it seems ok, but still `groupby-orderby-limit` need to look into since it cost more time somehow

The TSBS conducts a series of queries on the database, and our comparison of test results from versions 0.9 to 0.10 indicates some performance regression. Details are as follows:

![gtdb0 9vs0 10_p1](https://github.com/user-attachments/assets/d93f2666-8d59-4e33-9c2f-d1546bf1aa48)
![gtdb0 9vs0 10_p2](https://github.com/user-attachments/assets/8be7ab2f-663f-4a72-86bb-c82973abc90d)

As shown above, many of the GROUP BY queries are slower compared to version 0.9. The most significant regression is observed in the lastpoint query, which is five times slower in version 0.10 than in 0.9.

### Implementation challenges

We suspect that changes in the way sequential scans are handled between versions 0.9 and 0.10 may have caused these performance issues. Below are the metrics collected after repeatedly testing TSBS on version 0.10 of the database, obtained from `curl localhost:4000/metrics`:
[metrics.log](https://github.com/user-attachments/files/17914621/metrics.log)

**Comment by discord9** [comment]
we could set up a simple script running TSBS on a daily or weekly basis to detect such perf regression just like sqlite's speedtest?

**Comment by MichaelScofield** [comment]
@discord9 Just remember we have this https://github.com/GreptimeTeam/benchmark-scripts/blob/main/.github/workflows/bench.yml, and [it's still running](https://bencher.dev/console/projects/greptimedb/perf?key=true&reports_per_page=4&branches_per_page=8&testbeds_per_page=8&benchmarks_per_page=8&plots_per_page=8&reports_page=1&branches_page=1&testbeds_page=1&benchmarks_page=1&plots_page=1&branches=3b46708f-b87f-4f52-b1bb-1d9cc7bfee2d&heads=3b46708f-b87f-4f52-b1bb-1d9cc7bfee2d&testbeds=6d3be02f-9efe-4e47-8a5d-e389c228172d&benchmarks=da5c8cbe-9aef-431e-9168-11ef0821c8db%2Cbb7ce469-5c34-4a69-ab2f-d9769ca5be2a&measures=a2f1689d-44d5-4d5e-863f-47d285cedf97&start_time=1729902925000&end_time=1732494925000&lower_boundary=false&upper_boundary=false&clear=true)!

![image](https://github.com/user-attachments/assets/8173a209-e9ef-47a9-9ac6-a676088fcf22)

**Comment by evenyag** [comment]
I probably found two issues.

First, the SeqScan always pushes all row groups of a file to a range. Then the partition range has to merge all these row groups at the same time, which is unexpected.
https://github.com/GreptimeTeam/greptimedb/blob/8bdef776b3ef56c2ee8d2434cdb8ea3710aa01d3/src/mito2/src/read/range.rs#L266-L279

The performance regression is mainly caused by unbalanced partitions. In SeqScan, we split ranges by row groups if it only has one file. After partition assignment, the number of output partitions of SeqScan is equal to the target partition so the aggregation is done before repartition and the large partition becomes the bottleneck.
```
explain analyze SELECT max(usage_user), date_trunc('minute', ts) from cpu WHERE ts < '2023-06-13T13:26:46Z' group by date_trunc('minute', ts) order by date_trunc('minute', ts) desc limit 5;

|     0 |    0 |  MergeScanExec: peers=[4398046511104(1024, 0), ] metrics=[output_rows: 5, greptime_exec_read_cost: 0, finish_time: 6558674661, ready_time: 4747173, first_consume_time: 6557713984, ]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
|     1 |    0 |  GlobalLimitExec: skip=0, fetch=5 metrics=[output_rows: 5, elapsed_compute: 15359, ]
  SortPreservingMergeExec: [date_trunc(Utf8("minute"),cpu.ts)@1 DESC] metrics=[output_rows: 80, elapsed_compute: 55826, ]
    SortExec: TopK(fetch=5), expr=[date_trunc(Utf8("minute"),cpu.ts)@1 DESC], preserve_partitioning=[true] metrics=[output_rows: 80, elapsed_compute: 974578, row_replacements: 3464, ]
      ProjectionExec: expr=[MAX(cpu.usage_user)@1 as MAX(cpu.usage_user), date_trunc(Utf8("minute"),cpu.ts)@0 as date_trunc(Utf8("minute"),cpu.ts)] metrics=[output_rows: 3687, elapsed_compute: 15571, ]
        AggregateExec: mode=FinalPartitioned, gby=[date_trunc(Utf8("minute"),cpu.ts)@0 as date_trunc(Utf8("minute"),cpu.ts)], aggr=[MAX(cpu.usage_user)] metrics=[output_rows: 3687, elapsed_compute: 801764, ]
          CoalesceBatchesExec: target_batch_size=8192 metrics=[output_rows: 9945, elapsed_compute: 658692, ]
            RepartitionExec: partitioning=Hash([date_trunc(Utf8("minute"),cpu.ts)@0], 16), input_partitions=16 metrics=[send_time: 180563, repart_time: 749926, fetch_time: 7765834749, ]
              CoalesceBatchesExec: target_batch_size=8192 metrics=[output_rows: 9945, elapsed_compute: 87917, ]
                AggregateExec: mode=Partial, gby=[date_trunc(minute, ts@1) as date_trunc(Utf8("minute"),cpu.ts)], aggr=[MAX(cpu.usage_user)] metrics=[output_rows: 9945, elapsed_compute: 6265667736, ]
                  CoalesceBatchesExec: target_batch_size=8192 metrics=[output_rows: 88484000, elapsed_compute: 236376121, ]
                    FilterExec: ts@1 < 1686662806000000000 metrics=[output_rows: 88484000, elapsed_compute: 206774620, ]
                      SeqScan: region=4398046511104(1024, 0), partition_count=107 (0 memtable ranges, 6 file 110 ranges) metrics=[output_rows: 88484000, mem_used: 1775953664, elapsed_poll: 589859885, elapsed_await: 973069303, ]
 |
|  NULL | NULL | Total rows: 5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
3 rows in set (6.56 sec)
```

In v0.9.1, the partition number of SeqScan is less than the target partition so we have a repartition stage before aggregation.
```
|     0 |    0 |  MergeScanExec: peers=[4398046511104(1024, 0), ] metrics=[output_rows: 5, greptime_exec_read_cost: 0, first_consume_time: 1219146349, finish_time: 1219338274, ready_time: 2985165, ]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|     1 |    0 |  GlobalLimitExec: skip=0, fetch=5 metrics=[output_rows: 5, elapsed_compute: 8657, ]
  SortPreservingMergeExec: [date_trunc(Utf8("minute"),cpu.ts)@1 DESC] metrics=[output_rows: 3687, elapsed_compute: 161742, ]
    SortExec: expr=[date_trunc(Utf8("minute"),cpu.ts)@1 DESC], preserve_partitioning=[true] metrics=[output_rows: 3687, elapsed_compute: 16, spill_count: 0, spilled_bytes: 0, spilled_rows: 0, ]
      ProjectionExec: expr=[MAX(cpu.usage_user)@1 as MAX(cpu.usage_user), date_trunc(Utf8("minute"),cpu.ts)@0 as date_trunc(Utf8("minute"),cpu.ts)] metrics=[output_rows: 3687, elapsed_compute: 5783, ]
        AggregateExec: mode=FinalPartitioned, gby=[date_trunc(Utf8("minute"),cpu.ts)@0 as date_trunc(Utf8("minute"),cpu.ts)], aggr=[MAX(cpu.usage_user)] metrics=[output_rows: 3687, elapsed_compute: 936211, ]
          CoalesceBatchesExec: target_batch_size=8192 metrics=[output_rows: 58992, elapsed_compute: 427738, ]
            RepartitionExec: partitioning=Hash([date_trunc(Utf8("minute"),cpu.ts)@0], 16), input_partitions=16 metrics=[fetch_time: 19446510541, repart_time: 1491494, send_time: 144293, ]
              CoalesceBatchesExec: target_batch_size=8192 metrics=[output_rows: 58992, elapsed_compute: 2121242, ]
                AggregateExec: mode=Partial, gby=[date_trunc(minute, ts@1) as date_trunc(Utf8("minute"),cpu.ts)], aggr=[MAX(cpu.usage_user)] metrics=[output_rows: 58992, elapsed_compute: 5024914867, ]
                  CoalesceBatchesExec: target_batch_size=8192 metrics=[output_rows: 88484000, elapsed_compute: 270511984, ]
                    FilterExec: ts@1 < 1686662806000000000 metrics=[output_rows: 88484000, elapsed_compute: 214243640, ]
                      RepartitionExec: partitioning=RoundRobinBatch(16), input_partitions=4 metrics=[fetch_time: 1089026044, repart_time: 4, send_time: 72373349, ]
                        SeqScan: region=4398046511104(1024, 0), partition_count=4 (0 memtable ranges, 7 file ranges) metrics=[output_rows: 88484000, mem_used: 1775953664, elapsed_await: 1052567430, elapsed_poll: 447827426, ]
 |
|  NULL | NULL | Total rows: 5
3 rows in set (1.24 sec)
```

I'll send a PR to address this issue.

**Comment by discord9** [comment]
added query time benchmark to benchmark script: https://bencher.dev/console/projects/greptimedb/reports/73e606cd-6b5b-4e42-9f61-2b0c78469a7a?back=L2NvbnNvbGUvcHJvamVjdHMvZ3JlcHRpbWVkYi9yZXBvcnRzP3Blcl9wYWdlPTgmcGFnZT0x

**Comment by evenyag** [comment]
There is also a regression in compaction. The SeqScan uses row groups as sources to merge instead of files. The may cause compaction consume too much memory......
```
2024-12-05T09:22:58.985524Z  INFO mito2::read::seq_scan: DEBUG_SCAN: build all merge reader, num_files:
2, num_ranges: 1, num sources: 158
```

**Comment by evenyag** [comment]
fixed by #5090
